docs/design/node_agent_config-design.md

```diff
@@ -0,0 +1,91 @@
+- Deprecation of the "restic" configuration option
+- E2E tests for the `kopia` or `restic` uplaoder
+
+## High-Level Design
+
+Since new `nodeagentconfig` configuration option is a sibling of the `restic` one, the `restic` will be used as part of the new `nodeagentconfig` with addition of one `UploaderType` option that will be either `kopia` or `restic`.
+
+Without that option a default `restic` will be used. That default option may change in the future.
+
+## Detailed Design
+
+A new structure will be added, that includes the ResticConfig inline and extends this with one new parameter `UploaderType`:
+
+```go
+type NodeAgentConfig struct {
+	// Embedding ResticConfig
+	// +kubebuilder:validation:Optional
+	ResticConfig `json:",inline"`
+
+	// The type of uploader to transfer the data of pod volumes, the supported values are 'restic' or 'kopia'
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Enum=restic;kopia
+	UploaderType string `json:"uploaderType,omitempty"`
+}
+```
+
+The above `NodeAgentConfig` is a member of `ApplicationConfig`, which already includes `ResticConfig`, however we do not replace the `ResticConfig`, instead for backwards compatibility we add new `NodeAgentConfig` parameter:
+
+```go
+// ApplicationConfig defines the configuration for the Data Protection Application
+type ApplicationConfig struct {
+	Velero *VeleroConfig `json:"velero,omitempty"`
+	// (deprecation warning) ResticConfig is the configuration for restic server.
+	// Restic is for backwards compatibility and will be replaced by nodeagentconfig
+	// with the OADP 1.4+
+	// +optional
+	Restic *ResticConfig `json:"restic,omitempty"`
+
+	// NodeAgentConfig is needed to allow selection between kopia or restic
+	// +kubebuilder:validation:Optional
+	NodeAgent *NodeAgentConfig `json:"nodeagentconfig,omitempty"`
+}
+```
+
+## Alternatives Considered
+None.
+
+## Security Considerations
+
+The enablement of an extra upload mechanism could potentially introduce security implications due to the integration of a new backend, which might contain vulnerabilities.
+
+## Compatibility
+
+This design does not change current configuration options, however it proposes addition of the deprecation warning to the current `Restic` schema. The `Restic` configuration option will be removed from OADP 1.4+.
+
+## Implementation
+
+Implementation for the new data structure should be minimal.
+
+There are two uses of the additional `--uploader-type` that will specify restic or kopia uplaoder type.
+
+- One within call to the method from the `github.com/vmware-tanzu/velero` package:
+  ```go
+  func Deployment(namespace string, opts ...podTemplateOption) *appsv1.Deployment
+  ```
+
+- Second is within the OADP `pkg/velero/server/args.go` that is used for testing and development purposes.
+
+## Open Issues
+
+- Should we also rename the `restic.go` module to be named `node_agent.go` and relevant functions within the `restic.go`, so they are more generic.
```

Comment:
If restic.go is used for datamover/kopia as well, then we probably want to rename it.

Reply:
It's in the `controllers/restic.go` and `controllers/restic_test.go`:
  https://github.com/openshift/oadp-operator/blob/master/controllers/restic.go

I feel like it's better to rename it to:
`controllers/nodeagent.go` and  `controllers/nodeagent_test.go`

Reply:
CamelCase only for exported fields. snakeCase for private. File names is pretty flexible other than `*_test.go` I think